stonesoup/types/state.py

```diff
@@ -33,6 +34,136 @@ def ndim(self):
+
+        new_kwargs.update(kwargs)
+
+        return target_type(*args, **new_kwargs)
+
+
+class CreatableFromState:
+    class_mapping = {}
+
+    def __init_subclass__(cls, **kwargs):
+        bases = cls.__bases__
+        if CreatableFromState in bases:
+            # Direct subclasses should not be added to the class mapping, only subclasses of
+            # subclasses
+            return
+        if len(bases) != 2:
+            raise TypeError('A CreatableFromState subclass must have exactly two superclasses')
+        base_class, state_type = cls.__bases__
+        if not issubclass(base_class, CreatableFromState):
+            raise TypeError('The first superclass of a CreatableFromState subclass must be a '
+                            'CreatableFromState (or a subclass)')
+        if not issubclass(state_type, State):
+            # Non-state subclasses do not need adding to the class mapping, as they should not
+            # be created from States
+            return
+        if base_class not in CreatableFromState.class_mapping:
+            CreatableFromState.class_mapping[base_class] = {}
+        CreatableFromState.class_mapping[base_class][state_type] = cls
+        super().__init_subclass__(**kwargs)
+
+    @classmethod
+    def from_state(
+            cls,
+            state: State,
+            *args: Any,
+            target_type: Optional[type] = None,
+            **kwargs: Any) -> 'State':
+        """
+        Return new object instance of suitable type from an existing `state`.
+        The type returned can be explicitly specified using the `target_type` argument, otherwise
+        it is chosen by introspection of the created subclasses of this type: see below for an
+        example. Any compatible properties are copied from the input `state` to the returned
+        object, except for those specified by `args` and `kwargs`, which take precedence over those
+        from the input `state`.
+
+        This method is primarily concerned with type selection, with actual copying performed by
+        the static :meth:`~.State.from_state` method. As an example of the type selection
+        algorithm, consider the case of the class
+        `GaussianStatePrediction(Prediction, GaussianState)`. This is subclass of `Prediction`,
+        and `GaussianState` and so the `class_mapping` property will have an entry added (when
+        `GaussianStatePrediction` is defined) such that
+        `class_mapping[Prediction][GaussianState] = GaussianStatePrediction`. If this method is
+        then called like below
+
+        >>>> gaussian_state = GaussianState(some_arguments)
+        >>>> new_prediction = Prediction.from_state(gaussian_state, *args, **kwargs)
+
+        then the `from_state` method will look up the class mapping and see that
+        `Prediction.from_state()` called with a `GaussianState` input should return a
+        `GaussianStatePrediction` object, and therefore,the type of `new_prediction` will be
```

Comment:
"therefore,the type" -> "therefore, the type"

Reply:
Thanks. Should be addressed